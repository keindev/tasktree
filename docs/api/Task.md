# Task

Entity for managing a task - includes all child objects (informational messages, errors, progress bars, and tasks)

## Constructor

```typescript
new Task(text?: string, options?: ITaskOptions)
```

#### Parameters

| Name                | Type      | Description                                               |
| :------------------ | :-------- | :-------------------------------------------------------- |
| `text`              | _string_  | Task text message                                         |
| `options.autoClear` | _boolean_ | Removes all subtasks and progress bars after complete     |
| `options.status`    | _enum_    | New subtask status `(Completed, Failed, Pending,Skipped`) |

## Methods

### add

Create new subtask

### bar

Adds a new progress bar. Returns a progress bar object

### clear

Removes all subtasks and progress bars

### complete

Complete task

### error

Add new task error

### fail

Fail task

### log

Add log message

### render

Render task output

### skip

Skip task

### update

Update task text

### warn

Add task warning
