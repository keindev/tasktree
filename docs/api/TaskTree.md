# TaskTree

Singleton to manage the task tree

## Methods

### add

Adds a new task to the task tree. If there are active tasks, add a new one as a subtask - to the last subtask of the first active task

### exit

Force the process to exit (see process.exit). Do nothing in "silent mode"

### fail

Fail active task or adds a new subtask and call fail on it

### render

Render a task tree into a `string[]`. Returns strings with tasks hierarchy

### start

Starts output a task tree in a terminal at a defined interval. In “silent mode” - the task tree only collects tasks and is not output it in a terminal

### stop

Stop output a task tree in a terminal

### add

Adds a new task to the task tree. If there are active tasks, add a new one as a subtask - to the last subtask of the first active task

### fail

Fail active task or adds a new subtask and call fail on it

### tree

Method to get the object to control the tree (with [Theme](Theme.md))
