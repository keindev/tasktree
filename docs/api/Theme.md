# Theme

Allows you to set a custom theme options for the tree display.

## Options

All theme options is object with several fields:

```typescript
{
  // status text badge to the task, will be added after the task is completed
  badge?: string;
  // indication color
  color?: string;
  // indication symbol
  symbol?: string
}
```

### Indication types

| Type    | `badge`  | `color`                       |  `symbol`   | Description                                  |
| :------ | :------: | :---------------------------- | :---------: | :------------------------------------------- |
| default |    ✖     | `[default]` - text            |     `-`     | default color                                |
| active  |    ✖     | `#4285f4` - symbol            | `frame (⠧)` | spinner, progress bar color                  |
| success |    ✖     | `#00c851` - symbol, text, bar |      ✔      | task symbol, progress bar color              |
| skip    | `[skip]` | `#ff8800` - symbol, text, bar |      ↓      | task symbol, progress bar color              |
| error   | `[fail]` | `#ff4444` - symbol, text, bar |      ✖      | task symbol, error title, progress bar color |
| message |    ✖     | `#2e2e2e` - symbol            |      ─      | dim pointer to task information              |
| info    |    ✖     | `#33b5e5` - symbol            |      ℹ      | information message symbol                   |
| warning |    ✖     | `#ffbb33` - symbol            |      ⚠      | warning message symbol                       |
| subtask |    ✖     | `#2e2e2e` - symbol, text      |      ›      | dim pointer to subtask                       |
| list    |    ✖     | `#4285f4` - symbol            |      ❯      | list symbol                                  |
| dim     |    ✖     | `#838584` - symbol, bar       |     `-`     | dim color                                    |

> If you use a gradient fill for the progress bar - the color will change from `active` to `success`

```javascript
const theme = {
  default: '#ffffff',
  success: ['#008000', '✔'],
  skip: {
    symbol: '↓',
  },
  error: ['#ff0000', '✖', '[error]'],
  ...
};
```
