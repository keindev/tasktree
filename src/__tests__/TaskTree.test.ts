import stripAnsi from 'strip-ansi';
import { Terminal } from 'stdout-update/lib/terminal';
import { Task } from '../Task';
import { TaskTree, ITaskTreeOptions } from '../TaskTree';
import { Theme } from '../Theme';

const tree = TaskTree.tree();
const options: ITaskTreeOptions = { silent: true, autoClear: false };
const theme = new Theme();
const renderTree = (): string => stripAnsi(tree.render().join(Terminal.EOL));
const renderTask = (task: Task): string => stripAnsi(task.render(theme).join(Terminal.EOL));
let task: Task;

describe('TaskTree', (): void => {
    beforeEach((): void => {
        tree.start(options);
    });

    afterEach((): void => {
        tree.stop();
    });

    describe('Static', (): void => {
        it('Add', (): void => {
            TaskTree.add('task').complete();

            expect(renderTree()).toMatchSnapshot();
        });

        it('Fail', (): void => {
            try {
                tree.start(options);
                tree.fail('error message 1');
            } catch (error) {
                // eslint-disable-next-line jest/no-try-expect
                expect((error as Error).message).toBe('error message 1');

                tree.stop();
            }

            try {
                tree.start(options);
                tree.fail(new Error('error message 2'));
            } catch (error) {
                // eslint-disable-next-line jest/no-try-expect
                expect((error as Error).message).toBe('error message 2');

                tree.stop();
            }
        });
    });

    it('Default', (): void => {
        expect(tree).toBeDefined();
    });

    describe('Manage tasks', (): void => {
        beforeEach((): void => {
            task = tree.add('task');
            task.log(`message`).warn(`warning`);
        });

        it('Skip', (): void => {
            expect(renderTask(task.skip())).toMatchSnapshot();
        });

        it('Fail with string', (): void => {
            try {
                tree.fail('Something bad happened\nat X\nat Y\nat Z');
            } catch (error) {
                // eslint-disable-next-line jest/no-try-expect
                expect((error as Error).message).toBe('Something bad happened\nat X\nat Y\nat Z');
            }
        });

        it('Fail with new Error()', (): void => {
            try {
                tree.fail(new Error('Something bad happened\nat X\nat Y\nat Z'));
            } catch (error) {
                // eslint-disable-next-line jest/no-try-expect
                expect((error as Error).message).toBe('Something bad happened\nat X\nat Y\nat Z');
            }
        });

        it('Complete and render', (): void => {
            const output = renderTask(task.complete());

            expect(output).toBeTruthy();
            expect(output).toBe(renderTree());

            tree.stop();

            expect(renderTree()).toBe('');

            tree.start(options);
            tree.stop();

            expect(renderTree()).toBe('');
        });
    });
});
