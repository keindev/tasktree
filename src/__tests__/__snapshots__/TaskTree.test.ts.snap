// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskTree Manage tasks Skip 1`] = `
"  ↓ task [skip]
  ─ ⚠ warning
  ─ ℹ message"
`;

exports[`TaskTree Static Add 1`] = `"  ✔ task"`;
